$\qquad$ $\qquad$$\qquad$  **BBM409 Machine Learning: Assignmnet 2** <br />
$\qquad$ $\qquad$$\qquad$ **Goal: Naive Bayes**<br />
$\qquad$ $\qquad$$\qquad$                   **MELTEM TOKGÖZ** <br />
$\qquad$ $\qquad$$\qquad$                     **21527381** <br />
$\qquad$ $\qquad$$\qquad$                   **b21527381@cs.hacettepe.edu.tr** <br />

# PART I: Theoretical problems

MLE

1-Solution:  <img src ="mle1.png">

2-Solution:  <img src ="mle2.png">

3-Solution:  <img src ="mle3.png">


Naive Bayes

1-Solution: <img src ="nb1.png">

2-Solution: <img src ="nb2.png">

In [3]:
#********************************************************************************************************************************

# PART II: Detection of Fake News


In this part of the assignment, I will try to determine whether a headline is real or fake news.I will implement a Naive Bayes classifier.



Firstly, I used to import the libraries I used. I did use pandas, numpy, collections 

In [1]:
#**First-Step**Making imports******
import pandas as pd
import numpy as np
from collections import Counter

I used the nltk.corpus library for stopwords.

In [2]:
#stopword **********************************************
from nltk.corpus import stopwords
stopWords = set(stopwords.words('english'))

Then I opened the clean_fake-Train.txt file and clean_real-Train.txt file  given to me and started reading them.I read the test.csv with the pandas library.

In [3]:
#Read input file*****************************************
fake_input = open("clean_fake-Train.txt", "r")
real_input = open("clean_real-Train.txt","r")
test_input = pd.read_csv('test.csv', sep=',')
test_input.columns = ['id', 'category']

I created the dictionaries needed for unigram and bigram.

In [4]:
#UNIGRAM AND BIGRAM BAG OF WORDS IMPLEMENTATION*******************************************************

bow_all_u = {} #unigram all words dictionary
bow_ri_u = {}  #unigram all real input words in dictionary
bow_fi_u = {}  #unigram all fake input word in dictionary

bow_all_b = {} #bigram all words dictionary
bow_ri_b = {}  #bigram all real input words in dictionary
bow_fi_b = {}  #bigram all fake input word in dictionary

I've also created lists of words I'll use in the next step.I also created two int values to find the line numbers.

In [5]:
r_line=0
f_line=0

datafake =[] #fake data all line
datareal =[] #real data all line
words_counter = [] #all word

I read the real and fake file line by line and split it according to the gap and added it to the required dictionaries.I've used bag of word model and I added the words I read to the dictionary as follows: {word: total occurence word in file}
For example, x is 10 times in the real data {x: 10}.
I added the words in Unigram one by one to the dictionary.
I added two pairs in the bigram.I also looked at the words in the stop word.

In [6]:
for line in fake_input.readlines():
    f_line += 1
    line = line.rstrip("\n")
    datafake.append(line)
    line = line.split(" ")
    #unigram dictioary
    for words in line:
        #if words not in stopWords:
            words_counter.append(words)
            x = words
            if x not in bow_fi_u.keys():
                bow_fi_u[x] = 1
            else:
                bow_fi_u[x] += 1
            y = words
            if y not in bow_all_u.keys():
                bow_all_u[y] = 1
            else:
                bow_all_u[y] += 1
    #bigram dictionary
    for i in range(len(line)-1): 
        
        z = line[i],line[i+1]
        if z not in bow_fi_b.keys():
            bow_fi_b[z] = 1
        else:
            bow_fi_b[z] += 1

        t = line[i],line[i+1]
        if  t not in bow_all_b.keys():
            bow_all_b[t] = 1
        else:
            bow_all_b[t] += 1
    

for line in real_input.readlines():
    r_line += 1
    line = line.rstrip("\n")
    datareal.append(line)
    line = line.split(" ")
    #unigram dictionary
    for words in line:
        #if words not in stopWords:
                words_counter.append(words)
                x = words
                if x not in bow_ri_u.keys():
                    bow_ri_u[x] = 1
                else:
                    bow_ri_u[x] += 1 
        
                y = words
                if y not in bow_all_u.keys():
                    bow_all_u[y] = 1
                else:
                    bow_all_u[y] += 1
    #bigram dictionary
    for i in range(len(line)-1): 
        z = line[i],line[i+1]
        if  z not in bow_ri_b.keys():
            bow_ri_b[z] = 1
        else:
            bow_ri_b[z] += 1
 
        t = line[i],line[i+1] 
        if  t not in bow_all_b.keys():
            bow_all_b[t] = 1
        else:
            bow_all_b[t] += 1

In [7]:
print("Unigram Fake input unique word count : ",len(bow_fi_u))
print("Unigram Real input unique word count : ", len(bow_ri_u))
print("Unigram All unique word count : ", len(bow_all_u))

print("#######################################################")

print("Bigram Fake input unique word count : ",len(bow_fi_b))
print("Bigram Real input unique word count : ", len(bow_ri_b))
print("Bigram All unique word count : ", len(bow_all_b))

print("#######################################################")

Unigram Fake input unique word count :  3366
Unigram Real input unique word count :  3268
Unigram All unique word count :  5305
#######################################################
Bigram Fake input unique word count :  9164
Bigram Real input unique word count :  8589
Bigram All unique word count :  16801
#######################################################


1. Understanding the data
In this part I found the 3 words I asked for.Here I gave you 3 examples of real and fake  words along with statistics.

In [14]:
#highest 3 words reals unigram 
print("Stopwords not use")
print("#######################################################")
print("highest 3 words reals unigram :")
values1=list(bow_ri_u.values())
values1.sort()
last1=values1[-3:]
for i in bow_ri_u:
    if bow_ri_u[i] in last1:
        print(i, bow_ri_u[i])
print("#######################################################")
print("highest 3 words fake unigram :")
values2=list(bow_fi_u.values())
values2.sort()
last2=values2[-3:]
for i in bow_fi_u:
    if bow_fi_u[i] in last2:
        print(i, bow_fi_u[i])
print("#######################################################")

'''Stopword use'''
print("Stopwords use")
print("########################################################")
print("highest 3 words reals unigram :\ntrump 1484 \ndonald 709 \nus 186")
print("#######################################################")
print("highest 3 words fake unigram : \ntrump 1130 \ndonald 195 \nhillary 126")
#######################################################

Stopwords not use
#######################################################
highest 3 words reals unigram :
trump 1484
to 366
donald 709
#######################################################
highest 3 words fake unigram :
the 362
trump 1130
to 344
#######################################################
Stopwords use
########################################################
highest 3 words reals unigram :
trump 1484 
donald 709 
us 186
#######################################################
highest 3 words fake unigram : 
trump 1130 
donald 195 
hillary 126


I think these words are not exactly decisive because they are so much in both datasets.So ,I gave three more examples that I thought were more decisive.

<br>2. Implementing Naive Bayes  <br />
In this step I tried to classify each headline in the test file algorithm as fake or real with naive bayes algorithm.  <br />
<br>First of all, I applied this algorithm as in the example of chinese, which is described by our teacher.I used laplace smoothing because there might be words in the test file that are not in the dictionary.I first found the statistics that would be needed.

In [22]:
#Naïve Bayes Learning
#Prior Probability 
p_real =(r_line / (r_line + f_line))
p_fake =(f_line / (r_line + f_line))
print("P(real) = ",p_real)
print("P(fake) = ", p_fake)

#Naïve Bayes Classification
#Likelihood

total_r_uni = 0
for r in bow_ri_u.values():
    total_r_uni += r  
print("unigram total word real : " , total_r_uni)
    
total_f_uni = 0
for f in bow_fi_u.values():
    total_f_uni += f
print("unigram total word fake : " , total_f_uni)

print("#######################################################")

total_r_bi = 0
for r in bow_ri_b.values():
    total_r_bi += r  
print("bigram total word real : " , total_r_bi)
    
total_f_bi = 0
for f in bow_fi_b.values():
    total_f_bi += f
print("bigram total word fake : " , total_f_bi)

print("#######################################################")

#*****************************************************************************

P(real) =  0.602448685631977
P(fake) =  0.39755131436802305
unigram total word real :  13947
unigram total word fake :  13454
#######################################################
bigram total word real :  12274
bigram total word fake :  12350
#######################################################


I am reading the test file in the row row and I apply the following formula in the following algorithm.
<img src ="naivebayes.png">


In [24]:
p_real_dict_uni = {}
p_fake_dict_uni = {}

t_line = 0
uni_true_classify = 0
uni_false_classify = 0 
val1=[]

for i in range(0 , len(test_input)):
    p_r = 1
    p_f = 1
    t_line += 1
    test_line = test_input.values[i][0]
    for word in test_line.split(" "):
        #if word not in stopWords:
            a = word
            if a not in bow_ri_u.keys():
                p_real_dict_uni[a] = (0 + 1)/(total_r_uni + len(bow_all_u))
            else:
                p_real_dict_uni[a] = (bow_ri_u[a] + 1)/(total_r_uni + len(bow_all_u)) 
                    
            b = word
            if b not in bow_fi_u.keys():
                p_fake_dict_uni[b] = (0 + 1)/(total_f_uni + len(bow_all_u)) 
            else:
                p_fake_dict_uni[b] = (bow_fi_u[b] + 1)/(total_f_uni + len(bow_all_u))
                    
            c = word
            p_r  = (p_r * p_real_dict_uni[c])
            p_f  = (p_f * p_fake_dict_uni[c])
            
    p_r = (p_r * p_real)
    p_f = (p_f * p_fake)
    
    if p_f > p_r:
        val1.append('fake')  
    else:
        val1.append('real')
            
#********************************************************************************

print("total test number: ",t_line)

print("#######################################################")

num1 =0
for v in val1:
    if v == test_input.values[num1][1]:
        uni_true_classify += 1
    else:
        uni_false_classify += 1
    num1 += 1

print("Unigram true classify : ", uni_true_classify)
print("Unigram false classify : ", uni_false_classify)

print("#######################################################")

total test number:  489
#######################################################
Unigram true classify :  424
Unigram false classify :  65
#######################################################


I'm implementing the same algorithm for the bigram.Print results, you can see when you run.

In [25]:
p_real_dict_bi = {}
p_fake_dict_bi = {}

val2=[]

for i in range(0 , len(test_input)):
    p_r = 1
    p_f = 1
    test_line = test_input.values[i][0]
        
    line_b = test_line.split(" ")
    
    for i in range(len(line_b)-1): 
        
        a = line_b[i],line_b[i+1]
        if a not in bow_ri_b.keys():
            p_real_dict_bi[a] = (0 + 1)/(total_r_bi + len(bow_all_b))
        else:
            p_real_dict_bi[a] = (bow_ri_b[a] + 1)/(total_r_bi + len(bow_all_b)) 
       
        b = line_b[i],line_b[i+1]
        if b not in bow_fi_b.keys():
            p_fake_dict_bi[b] = (0 + 1)/(total_f_bi + len(bow_all_b))
        else:
            p_fake_dict_bi[b] = (bow_fi_b[b] + 1)/(total_f_bi + len(bow_all_b)) 
        
        c = line_b[i],line_b[i+1]
        p_r  = (p_r * p_real_dict_bi[c])
        p_f = (p_f * p_fake_dict_bi[c])
 
    p_r = np.log10(p_r * p_real)
    p_f = np.log10(p_f * p_fake)
    
    if p_f > p_r:
        val2.append('fake')
    else:
        val2.append('real')
        
#print("#######################################################")

bi_true_classify = 0
bi_false_classify = 0
num2 =0
for v in val2:
    if v == test_input.values[num2][1]:
        bi_true_classify += 1
    else:
        bi_false_classify += 1
    num2 += 1
    
print("bigram true classify : ", bi_true_classify)
print("bigram false classify : ", bi_false_classify)

print("#######################################################")

bigram true classify :  407
bigram false classify :  82
#######################################################


Because Bigram also looked at the relationship between words, he had to make more accurate predictions, but he did not.1-2 points more accurate.I think this is due to the small size of the data set.

Finally, I calculated the accuracy.When I'm not using stopword, accuracy is like this:

In [26]:
accuracy_uni = 100 * (uni_true_classify / t_line)
print("unigram Accuracy : ",accuracy_uni)

accuracy_bi = 100 * (bi_true_classify / t_line)
print("bigram Accuracy : ",accuracy_bi)

unigram Accuracy :  86.70756646216769
bigram Accuracy :  83.23108384458078


When I'm using stopword, accuracy is like this:

In [28]:
print("unigram Accuracy :  85.2760736196319")
print("bigram Accuracy :  83.23108384458078")

unigram Accuracy :  85.2760736196319
bigram Accuracy :  83.23108384458078


Stoword aims to filter the most commonly used words.
Therefore, it may be useful to use it because it needs to increase accuracy.
I have observed a few points drop accuracy  when I use the stopword.
I don't know exactly why, but as I said earlier, we might be testing on a small data set.

(a) Analyzing effect of the words on prediction

In this part, I have calculated the desired words as follows.
<img src ="tenwords.png">

In [29]:
print("################PART 3 ###################################\n")

bag_of_word = Counter(words_counter)
all_lines = datafake + datareal

'''
P(class) = p_real , p_fake
P(word|class) = #of occurance of words in headlines of that class / total number of headlines of that class
'''
real_prior = np.log10(p_real)
fake_prior = np.log10(p_fake)

pword_bow = {}
for w in bag_of_word.keys():
    
    hl_num_f = 1
    for l in datafake:
        if w in l:
            hl_num_f += 1
    w_fake = hl_num_f / len(datafake)
    
    hl_num_r = 1
    for l in datareal:
        if w in l:
            hl_num_r += 1
    w_real = hl_num_r / len(datareal)

    pword_bow[w] = [w_real, w_fake]

'''
#PRESENCE

P(class|word) = P(word|class) * P(class) / (P(word|fake)*P(fake) + P(word|real)*P(real))

'''
real_w = []
fake_w = []

for w in pword_bow.keys():
    val = pword_bow[w]
    real_w.append([w, val[0] * p_real / (val[0] * p_real + val[1] * p_fake)])
    fake_w.append([w, val[1] * p_fake / (val[0] * p_real + val[1] * p_fake)])

print("*******Presence for real***********")
for i in range(10):
    print(sorted(real_w, key=lambda tup: tup[1], reverse=True)[i])
    
print("******Presence for fake************")
for i in range(10):
    print(sorted(fake_w, key=lambda tup: tup[1], reverse=True)[i])

'''
#ABSENCE

P(~word|class) = 1- P(word|class)
P(~word) = headlines without word / # headlines

P(class|~word) = P(~word|class)*P(class)/P(~word)

'''
real_not_w = []
fake_not_w = []

for w in pword_bow.keys():
    headline_num = 1
    for l in all_lines:
        if not w in l:
            headline_num += 1
    not_word = headline_num / len(all_lines)
    val = pword_bow[w]
    fake_not_w.append([w, (1 - val[1]) * p_fake / not_word])
    real_not_w.append([w, (1 - val[0]) * p_real / not_word])

print("****Absence for real******")
for i in range(10):
    print(sorted(real_not_w, key=lambda tup: tup[1], reverse=True)[i])

print("****Absence for fake******")
for i in range(10):
    print(sorted(fake_not_w, key=lambda tup: tup[1], reverse=True)[i])


*******Presence for real***********
['korea', 0.986111111111111]
['turnbull', 0.9807692307692307]
['travel', 0.9795918367346939]
['australia', 0.9777777777777777]
['trumps', 0.9736842105263158]
['climate', 0.9615384615384615]
['north', 0.9583333333333333]
['paris', 0.9545454545454546]
['flynn', 0.95]
['asia', 0.9444444444444444]
******Presence for fake************
['breaking', 0.9615384615384615]
['soros', 0.95]
['black', 0.9487179487179487]
['reason', 0.9473684210526315]
['woman', 0.9411764705882353]
['steal', 0.9285714285714285]
['interview', 0.923076923076923]
['homeless', 0.9090909090909091]
['12', 0.9090909090909091]
['endingfed', 0.9090909090909091]
****Absence for real******
['i', 0.7959183673469389]
['y', 0.7118133935289691]
['c', 0.6976744186046511]
['in', 0.6958970233306517]
['f', 0.6934594168636723]
['he', 0.691827173347215]
['w', 0.6863672182821118]
['l', 0.6794425087108011]
['er', 0.6729475100942126]
['the', 0.6696548632900045]
****Absence for fake******
['al', 0.493150684

Especially when I use the stopword, I see that the real absences and fake presence are similar.Likekly, fake absences and real presence is similar.Because 

<br>****Absence for real******<br\>
<br>['c', 0.6976744186046511]<br\>
<br>['f', 0.6934594168636723]<br\>
<br>['w', 0.6863672182821118]<br\>
<br>['l', 0.6794425087108011]<br\>
<br>['er', 0.6729475100942126]<br\>
<br>['v', 0.6643109540636043]<br\>
<br>['b', 0.663359319631467]<br\>
<br>['nt', 0.6532426778242677]<br\>
<br>['il', 0.6519133241124943]<br\>
<br>['ed', 0.6496964035497431]<br\>
<br>****Absence for fake******<br\>
<br>['al', 0.4931506849315068]<br\>
<br>['na', 0.4893238434163701]<br\>
<br>['donald', 0.483974358974359]<br\>
<br>['n', 0.42574257425742584]<br\>
<br>['trumps', 0.4243243243243243]<br\>
<br>['says', 0.41169635941130905]<br\>
<br>['say', 0.41100702576112413]<br\>
<br>['ban', 0.40733807562710594]<br\>
<br>['korea', 0.40731166912850814]<br\>
<br>['north', 0.4065731166912851]<br\>
<br>*******Presence for real***********<br\>
<br>['korea', 0.986111111111111]<br\>
<br>['turnbull', 0.9807692307692307]<br\>
<br>['travel', 0.9795918367346939]<br\>
<br>['australia', 0.9777777777777777]<br\>
<br>['trumps', 0.9736842105263158]<br\>
<br>['climate', 0.9615384615384615]<br\>
<br>['north', 0.9583333333333333]<br\>
<br>['paris', 0.9545454545454546]<br\>
<br>['flynn', 0.95]<br\>
<br>['asia', 0.9444444444444444]<br\>
<br>******Presence for fake************<br\>
<br>['breaking', 0.9615384615384615]<br\>
<br>['soros', 0.95]<br\>
<br>['black', 0.9487179487179487]<br\>
<br>['reason', 0.9473684210526315]<br\>
<br>['woman', 0.9411764705882353]<br\>
<br>['steal', 0.9285714285714285]<br\>
<br>['interview', 0.923076923076923]<br\>
<br>['endingfed', 0.9090909090909091]<br\>
<br>['homeless', 0.9090909090909091]<br\>
<br>['12', 0.9090909090909091]<br\>


<br>The result is that when I remove the words presence and make naive bayes, the result is like this.<br\>

<br>unigram Accuracy :  86.29856850715747<br\>
<br>bigram Accuracy :  83.23108384458078<br\>
    
<br>The result is that when I remove the words absence and make naive bayes, the result is like this.<br\>

<br>unigram Accuracy :  86.29856850715747<br\>
<br>bigram Accuracy :  83.23108384458078<br\>

This assignment best accuracy is unigram Accuracy :  86.70756646216769